cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/DatasetServiceClient.java

```diff
@@ -304,9 +304,15 @@ private HttpResponse doDelete(String resource)
   private HttpResponse doRequest(HttpRequest.Builder requestBuilder)
     throws DatasetManagementException, UnauthorizedException {
     HttpRequest request = addUserIdHeader(requestBuilder).build();
+    HttpResponse response;
     try {
       LOG.trace("Executing {} {}", request.getMethod(), request.getURL().getPath());
-      HttpResponse response = remoteClient.execute(request);
+      HttpMethod method = request.getMethod();
+      if (method == HttpMethod.GET || method == HttpMethod.PUT || method == HttpMethod.DELETE) {
```

Comment:
Why this logic can't be inside the `RemoteClient` so that it applies to all usages?

Reply:
A good suggestion. Can we have an enum, e.g. 
```
enum Idempotency { NONE, IDEMPOTENT, AUTO }
```
and instead of `executeIdempotent` have `execute(HttpRequest request, Idempotency idempotency)` that overloads regular execute and does a method check on AUTO

Reply:
This way here we will just call `remoteClient.execute(request, Idempotency.AUTO)`

Reply:
PUT and DELETE may not be Idempotent always. But Ankit probably thought for DatasetServiceClient, based on current usage, these are safe to retry.

Maybe we should have some notion of a unique operationid. If operationid has been seen by the service it ignores. Essentially, the service takes responsibility of handling Idempotence. The client always retries.

Reply:
Correction. https://restfulapi.net/idempotent-rest-apis/ looks like PUT and DELETE were meant to be Idempotent. In that case, Terence'a suggestion to move them into RemoteClient makes sense. @tivv why do we need the the enum? Is it to handle cases where we know some methods are idempotent even if they are a POST (not GET, PUT, DELETE).

Reply:
We have 3 cases:
 - Caller knows the call is idempotent irrespective of method
 - Caller knows the call is not idempotent irrespective of method
 - Caller wants to use "default" logic - call is idempotent for GET/PUT/DELETE and is not idempotent for other methods